### DISCUSSION TITLE
Create a custom node that wraps existing node

### DISCUSSION THREAD

**Discussion by ian-whitestone** [markdoc/markdoc#246] [Q&A] [answered]
# Background

Using markdoc for our websites blog & documentation.

The built-in table node works great on desktop, but on mobile it does not overflow.

https://select.dev/posts/cost-per-query#sql-implementation-2

## Desktop

<img width="1264" alt="CleanShot 2022-10-11 at 08 19 45@2x" src="https://user-images.githubusercontent.com/11096727/195087884-265259d4-5f6a-462c-b1ff-c8ed83df30b4.png">

## Mobile

<img width="502" alt="CleanShot 2022-10-11 at 08 20 15@2x" src="https://user-images.githubusercontent.com/11096727/195087932-182b55cf-ff24-4c01-a7dd-22dd93c188d3.png">

# Proposed Fix

If I wrap the `<table>` with:

```html
<div style="overflow-x:auto;">
  <table>
    ...
  </table>
</div>
```

Then it just becomes a scrollable container:

![CleanShot 2022-10-11 at 08 24 33](https://user-images.githubusercontent.com/11096727/195089051-464253f3-718d-4142-af90-2b254838a4c5.gif)

# Question

To accomplish this, I believe I need to create a custom table node that overrides the default one. But I really don't want to have to re-write all the table parsing/rendering functionality, I just want to wrap the existing functionality with a `div`, i.e. something like:

```js
// components/CustomTable.js
export function CustomTable({ ... }) {
    return (
      <div style="overflow-x:auto;">
             <ExistingMarkdocTable content={content} />
      </div>
   )
}
```

Is this possible? If so, able to share an example 🙏 ?

Thanks!

**Comment by duncanma-stripe** [comment]
How about just adding 

```
article table {
    overflow-x: auto;
}
```

to your css file?

**Reply by ian-whitestone** [reply]
Thanks @duncanma-stripe. CSS noob here: what does this effectively do? Does it nest all tables in an `article` element? I'd only like to apply this to tables in the blog, not necessarily site wide.

**Reply by duncanma-stripe** [reply]
So it doesn't change the markup of the site at all, it just says "add these styles to all tables that are inside an `article` element"

**Reply by duncanma-stripe** [reply]
If you are using the `{% table %}` syntax, you could have even more control by doing 

```

{% table .scrollable %}
 - line 1
 - line 2
{% /table %}

```

and then having this in your css

```
table.scrollable {
    overflow-x: auto;
}
```

So only tables where you put .scrollable would be impacted.

**Reply by ian-whitestone** [reply]
Thanks @duncanma-stripe ! really appreciate the additional comments. It seems like the `overflow-x` style doesn't work when applied to a table element,  so I think i'll need to need to wrap it in a `div` and apply the styling there. Will try out @rpaul-stripe's solution!

**Comment by rpaul-stripe** [comment, marked as answer]
Duncan's solution is the best approach for your use case. But for completeness, this is how you would override the built-in table node to wrap its output in a div:

```
const Markdoc = require('@markdoc/markdoc');

const config = {
  nodes: {
    table: {
      ...Markdoc.nodes.table,

      transform(node, config) {
        const attrs = node.transformAttributes(config);
        const children = node.transformChildren(config);

        return new Markdoc.Tag('div', { class: 'table-container' }, [
          new Markdoc.Tag('table', attrs, children),
        ]);
      },
    },
  },
};
```

**Reply by ian-whitestone** [reply]
Thanks @rpaul-stripe , I ended up using this approach (see https://github.com/markdoc/markdoc/discussions/246#discussioncomment-3854573 for why) and it worked.

One thing to note for any future readers who are using TailwindCSS, I first tried putting a TailwindCSS class (`overflow-x-auto`, which is the same as setting `style="overflow-x:auto;"`) directly in the `transform` function instead of creating a new CSS class, i.e.

```javascript
return new Markdoc.Tag('div', { class: 'overflow-x-auto' }, [
          new Markdoc.Tag('table', attrs, children),
        ]);
```

But it did not work, for whatever reason. Wild guess - maybe the tailwind CSS build happens at a different time, or doesn't pick up that this new class is being used?

**Reply by benrudolphWSP** [reply]
Late to the party, but for future readers. Tailwind most likely is not picking up the class because you need to tell Tailwind where to look for classes. In your config file.

```javascript
/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ['./src/**/*.{js,jsx}', './markdoc/**/*.{js,jsx}'],
  ...
}
```

**Reply by benrudolphWSP** [reply]
What if I wanted to inject a caption into the table? I'd like to make my tables a little more accessible with captions. Is this possible? Could this be accomplished with slots (https://github.com/markdoc/markdoc/discussions/342) possibly?

**Reply by rpaul-stripe** [reply]
@benrudolphWSP yes, here's a simple example that demonstrates how to do this with slots: https://gist.github.com/rpaul-stripe/fb0a51c41baec931cb36e3301d6765b2
